### ISSUE TITLE
Set custom comment for each archive created by the Streamer

### ISSUE THREAD

**Issue by RTrampov** [WeTransfer/zip_tricks#40]
Hi guys and thanks for the great library.

I recently implemented support for streaming zip downloads in a project I'm working on and there were two features that we required that weren't available out of the box. I will open separate issues for both of them so they can be tracked/discussed separately.

The first one is the the ability to write a comment for each created archive by the `ZipTricks::Streamer` - an example usage would be:
```ruby
ZipTricks::Streamer.open(output_stream) do |streamer|
  # add some files

  streamer.zip_comment = 'Some comment'
end
```

If you are interested in such functionality I can open a pull request.

**Comment by julik** [comment]
Strictly speaking there is a possibility to do this now already by overriding the writer, which is injectable.

```
class CustomWriter < ZipTricks::ZipWriter
  def write_end_of_central_directory(**kwargs)
    super(**kwargs, comment: '')
  end
end

ZipTricks::Streamer.open(io, writer: CustomWriter.new) do |z|
  # ...
end
```

I know that having a method for this is slightly nicer but this adds more state the Streamer object has to carry around, which is not ideal. For your particular case, would a writer override work?

**Comment by RTrampov** [comment]
Yeah, that's what we tried initially but we prefer not to patch the libraries we use. If for some reason the method signature of `write_end_of_central_directory` changes in the future the custom writer would stop working.

My idea was that the streamer already calls `write_end_of_central_directory` which takes `comment` as an argument, so it's the most obvious place to pass it.

Another solution would be to make the comment configurable in the writer's initializer like so:
```ruby
class ZipTricks::ZipWriter
  def initialize(comment = ZIP_TRICKS_COMMENT)
    @comment = comment
  end

  def write_end_of_central_directory(io:, start_of_central_directory_location:, central_directory_size:, num_files_in_archive:, comment: @comment)
    # ...
  end
end
```

And then it can be used when opening new streamer:
```ruby
ZipTricks::Streamer.open(io, writer: ZipTricks::ZipWriter.new('some comment') do |streamer|
  # ...
end
```

**Comment by julik** [comment]
> Yeah, that's what we tried initially but we prefer not to patch the libraries we use.

I relate but the method you are overriding in this case is a public API. You are not patching it - you are extending it, and from the maintainer perspective I can tell you that we do pay attention to semver. I.e. as long as we are on the same major version of ZipTricks the overridden method will work.

> Another solution would be to make the comment configurable in the writer's initializer like so:

This is a great thing to do in a custom writer object. We tried to make the Writer completely stateless so that it doesn't impose any constraints on the caller and can potentially be reused across multiple calls.

In summary: I really suggest you try extending the Writer for your specific case instead, in the application. Not because this feature is impossible to have, but because I believe it cuts a bit too much "upwards" - and since there is an existing public API just for that ;-)
